x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/IndexLifecycleExplainResponse.java

```diff
@@ -227,9 +227,22 @@ private IndexLifecycleExplainResponse(
             if (policyName == null) {
                 throw new IllegalArgumentException("[" + POLICY_NAME_FIELD.getPreferredName() + "] cannot be null for managed index");
             }
-            // check to make sure that step details are either all null or all set.
-            long numNull = Stream.of(phase, action, step).filter(Objects::isNull).count();
-            if (numNull > 0 && numNull < 3) {
+
+            // check to make sure that the required step details are either all null or all set.
+            final long maxNull;
+            final Stream<String> conditions;
+
+            // step==ERROR allows phase,action to be set or null
+            if (ErrorStep.NAME.equals(step)) {
+                maxNull = 2;
+                conditions = Stream.of(phase, action);
+            } else {
+                maxNull = 3;
+                conditions = Stream.of(phase, action, step);
+            }
+
+            long numNull = conditions.filter(Objects::isNull).count();
+            if (numNull > 0 && numNull < maxNull) {
```

Comment:
I think this method might now be too clever; it's hard to reason about. Can we make it simpler? Perhaps something like:

```java
// If at least one detail is null, but not *all* are null
if (Stream.of(phase, action, step).anyMatch(Objects::isNull) && Stream.of(phase, action, step).allMatch(Objects::isNull) == false) {
    // …and it's not in the error step
    if (ErrorStep.NAME.equals(step) == false) {
        // then throw an exception about missing the complete step details
        throw new IllegalArgumentException("…");
    }
}
```

Reply:
Way more clear--thanks!